### PR TITLE
#15 better support for compiler plugins

### PR DESCRIPTION
**Pull request by Foso**
This PR will added the option to add a list of ComponentRegistrars to the compiler.

### CODE REVIEW

**Comment by tschuchortdev** [comment]
Looks nice. Arrow-meta was announced to the public at lambda world conf this Friday, so it's good if we have more ergonomic support for compiler plugins soon when more people will write their own. 

Is this ready to be merged or more like a WIP? I will do a closer review tomorrow. 

The stupid CI failed again but I think it's unrelated to these changes.

**Comment by Foso** [comment]
Hi, i heard about arrow-meta and i'm excited to try it out :) 

This is ready to be merged, i wanted to add this feature, before i will add the support of the JSCompiler

File: src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt [hunk 1]
```diff
@@ -409,10 +414,6 @@ class KotlinCompilation {
 	}
 	/** Performs the 1st and 2nd compilation step to generate stubs and run annotation processors */
 	private fun stubsAndApt(sourceFiles: List<File>): ExitCode {
-		if(annotationProcessors.isEmpty()) {
-			log("No services were given. Not running kapt steps.")
-			return ExitCode.OK
-		}
```
[the remark this answers is not in the document]
Comment (Foso) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: I removed this because it was not possible to execute the compiler plugins without also adding an annotation processor.
The problem was that the pluginClasspaths were set after this check.

I restored this and made changes. Inside stubsAndApt, only the annotation processors are added to threadLocalParameters. Then inside compileKotlin, only the compiler plugins are set, so the processors aren't getting executed twice.

File: src/test/kotlin/com/tschuchort/compiletesting/CompilerPluginsTest.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import org.junit.Test
+import org.mockito.Mockito
+import javax.annotation.processing.AbstractProcessor
+import javax.annotation.processing.RoundEnvironment
+import javax.lang.model.element.TypeElement
+
+class CompilerPluginsTest{
+
+    @Test
+    fun whenCompilerPluginsAreAddedTheyGetExecuted() {
```
[the remark this answers is not in the document]
Comment (Foso) [reply src/test/kotlin/com/tschuchort/compiletesting/CompilerPluginsTest.kt]: I've renamed the tests

File: src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (tschuchortdev) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: Wouldn't this change the behaviour so that additional compiler plugins are only registered in the `compileKotlin` step? The current behaviour of #23. Is that additional plugins are executed in both steps. I don't know which way is actually right though,

[the remark this answers is not in the document]
Comment (Foso) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: Maybe you are right, we could also do it this way:

```kotlin
if(annotationProcessors.isEmpty() && compilerPlugins.isEmpty()) {
			log("No services were given. Not running kapt steps.")
			return ExitCode.OK
}
```
I remove the changes in stubsAndApt and compileKotlin and just add the additional check for the compilerPlugins. Then the behaviour stays the same and it's possible to add compilerPlugins without also adding annotation processors.

[the remark this answers is not in the document]
Comment (tschuchortdev) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: I think it would be best to always add compiler plugins in the `stubsAndApt` step but cancel it if there are no annotation processors, like we did previously.

File: src/main/kotlin/com/tschuchort/compiletesting/MainComponentRegistrar.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import org.jetbrains.kotlin.com.intellij.mock.MockProject
+import org.jetbrains.kotlin.compiler.plugin.ComponentRegistrar
+import org.jetbrains.kotlin.config.CompilerConfiguration
+import org.jetbrains.kotlin.kapt3.base.incremental.IncrementalProcessor
+
+internal class MainComponentRegistrar : ComponentRegistrar {
+
+    override fun registerProjectComponents(project: MockProject, configuration: CompilerConfiguration) {
+        val parameters = threadLocalParameters.get()
+        KaptComponentRegistrar(parameters.processors,parameters.kaptOptions).registerProjectComponents(project,configuration)
```
Comment (tschuchortdev) [inline src/main/kotlin/com/tschuchort/compiletesting/MainComponentRegistrar.kt]: I read a few days ago in the Kotlin slack that the order of registering compiler plugins matters. Right now we are registering Kapt first. Could there be a problem here? For example if a compiler plugin like arrow-meta introduces some new syntax/type checking that isn't legal in vanilla Kotlin the respective plugin would have to be run before Kapt or Kapt wouldn't know how to deal with that. Any thoughts on this @raulraja?

**Comment by raulraja** [comment]
Unless I'm missing this is fine because meta does not rely on java stubs,
it's interleaved with the compiler. But you may want to offer kapt
registration as opt in and let them specify which plugin they want to run
supporting a predefined set of known plugins like kapt, serialization, meta
etc.

**Comment by tschuchortdev** [comment]
@raulraja That's not really what I meant. Let's assume you have some plugin that replaces every method named `foo` with a method `bar`. If the Kapt plugin is registered first, I assume that apt will see a `foo` method in the stub files. If Kapt is registered last, then I assume apt will see a `bar` method in the stub files. This could be problematic.

**Comment by raulraja** [comment]
Yes, I see what you mean now, sorry I misunderstood. Yes, in that case
order matters.

File: src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Foso) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: I'm not sure if i understand it correctly, in stubsAndApt we'll add the compiler plugins and cancel if there is no annotation processor. And then in compileKotlin we'll add the plugins again?

Otherwise i can't see how it's possible to add a compiler plugin without also having to add a annotation processor.

File: src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt [hunk 4]
```diff
@@ -523,22 +523,6 @@ class KotlinCompilation {
...
-		 */
-		MainComponentRegistrar.threadLocalParameters.set(
-			MainComponentRegistrar.Parameters(
-				listOf(),
-				KaptOptions.Builder(),
-				compilerPlugins
-			)
-		)
-
-
```
Comment (tschuchortdev) [inline src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: Why did you remove this again? I think we always need to add the compiler plugins to the `compileKotlin` step, even if they already ran in the `stubsAndApt` step.

**Comment by tschuchortdev** [comment]
@Foso Can you check what the correct behaviour is when KApt interacts with compiler plugins? If you just modify your MpApt plugin to add some dummy method to all classes and then run it together with a regular annotation processor, you should be able to see the method (or not) in the AP and we can be sure. I'm not so familiar with compiler plugins unfortunately or I would do it myself. 

If not, I say we just register first-party plugins before KApt in the `MainComponentRegistrar` and merge it. If someone complains we can always fix it later.

File: src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt (hunk 4 above)
Comment (Foso) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: H @tschuchortdev , i removed it because i wanted to change a less as possible, because i couldn't guarantee that it would break any behaviour of any other compiler plugins that runs in any of the compiler phases. All i can say is, that i could test my MpApt Plugin with these changes and it didn't break any of the existing tests and Kapt was also working. I don't know much about when which compiler phase is running, because my plugin is only running in one phase.

Comment (tschuchortdev) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: Are you creating new source files with your plugin or simply adding to the AST?

Comment (Foso) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: I'm creating new source files

Comment (Foso) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: My plugins just detects the annotations, it doesn't provide code generation

Comment (tschuchortdev) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: I think we still need to run the plugins in the `compileKotlin` step then, because all the changes done to the AST during `stubsAndApt` will be discarded, so a compiler plugin that changes the AST would have no effect.  

But I think we also need to run the plugins in the `stubsAndApt` step as well, before kapt plugin is run.

Comment (Foso) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: I also think that we should add it to both. I could restore my deleted version.

Comment (tschuchortdev) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: Sounds good 👍

Comment (Foso) [reply src/main/kotlin/com/tschuchort/compiletesting/KotlinCompilation.kt]: I restored it, and changed the check in stubsAndApt to "annotationProcessors.isEmpty()" again, because without, all added plugins were executed in both phases. In the future we need an option where you can set in which phases you want to execute the plugins, but for now i would keep it that way. If you think otherwise, just add " && compilerPlugins.isEmpty()" to the if statement again.
